### PR TITLE
CMake: Use toolchain files

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: scripts/travis-emscripten/build_emscripten.sh
Comment (axic) [inline scripts/travis-emscripten/build_emscripten.sh:84]: If this is not set anymore, will the `elsif(EMSCRIPTEN)` in `EthCompilerSettings` still work?

Comment (chfast) [reply scripts/travis-emscripten/build_emscripten.sh:84]: This is set by the emscripten toolchain file from Emscripten SDK. So it not works like `if(APPLE)` or `if(MSVC)`.

**Review by chriseth — DISMISSED** [review]

**Comment by chfast** [comment]
Wait with this. I see a lot of failures.

**Comment by axic** [comment]
@chfast are you still working on this?

**Comment by chfast** [comment]
Not actively.

I have problem with Emscripten - I'm not able to mix C++11 toolchain with Emscripten toolchain. Everytime I touch something related to Emscripten the Travis build fails without enough information in logs... I will go back to this some day, but probably not this week.

**Comment by axic** [comment]
@chfast was this problem resolved?

**Comment by chfast** [comment]
No. Emscripten toolchain is not fully supported by Hunter and you cannot build Boost this way.

**Comment by axic** [comment]
@chfast what should we do with this PR in that case?

**Comment by axic** [comment]
This should help by using the sane default from Emscripten via the toolchain file.

@chfast can this be finished somehow?

**Comment by chfast** [comment]
Yes, the idea it's simple: you have 2 toolchain files: one for regular C++11 builds, one for Emscripten builds.

But this not work in practice, because Hunter does not support building boost with Emscripten this way. It requires some work there, probably quite time consuming.

**Comment by axic** [comment]
@chfast are you saying this PR should be abandoned or not?

If it should be abandoned, does it still make sense creating an issue and linking here so that someone in the future might pick it up?

**Comment by axic** [comment]
@ekpyron here's another CMake PR to think about.

**Comment by ekpyron** [comment]
@axic Yes - I like this one. I think we should decide about #4486 first, though - it should make things easier.

**Comment by axic** [comment]
@ekpyron are you still working on this?

**Comment by ekpyron** [comment]
@axic Not right now, but looking into this and generally improving the build system is on my list.

**Comment by axic** [comment]
Rebased.

**Comment by axic** [comment]
@ekpyron @chfast wow this actually works after we've merged the emscripten updates.

File: .travis.yml
Comment (chfast) [inline .travis.yml]: Should it stay?

Comment (axic) [reply .travis.yml]: No, all those build triggering changes will have to go.

**Comment by chriseth** [comment]
Moving from "needs review" to "in progress" as the title says "WIP".

**Comment by axic** [comment]
> Moving from "needs review" to "in progress" as the title says "WIP".

The title wrongfully said it [after](https://github.com/ethereum/solidity/pull/2836#issuecomment-440762815). It needs to be reviewed.

**Comment by axic** [comment]
Rebased this and updated to C++14, not sure if it works anymore.

**Comment by chriseth** [comment]
What is this about? What is the benefit?

**Comment by axic** [comment]
I think the main benefit is to properly (aka cmake style) set the compiler version. @chfast can answer better probably.

**Comment by chriseth** [comment]
Seems to be fine.

**Review by chriseth — APPROVED** [review]
Looks good, but we should also change the

**Comment by chfast** [comment]
This is a way to extend some CMake settings to dependencies. Hunter relay on this. E.g. you can have a toolchain file that configures the build for Emscripten and "magically" whole project will cross-compile with Emscripten.
But this allows you also do something simpler without adding additional options to project, e.g. compile in C++14 mode.

There are my example toolchains: https://github.com/ethereum/cable/tree/master/toolchains
I also have a `-march=native` somewhere for benchmarking.

Here are more: https://github.com/ruslo/polly.

**Pull request merged**
